dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+        Func<T, Task<TResult>> workflow, T arg, StartWorkflowOptions options) { /*...*/ }
+
+    public async Task<WorkflowHandle<Nothing>> StartWorkflowAsync(
+        string workflow, object[] args, StartWorkflowOptions options) { /*...*/ }
+
+    public async Task<WorkflowHandle<TResult>> StartWorkflowAsync<TResult>(
+        string workflow, object[] args, StartWorkflowOptions options) { /*...*/ }
+
+    // GetWorkflowHandle
+
+    public WorkflowHandle<Nothing> GetWorkflowHandle(
+        string workflowID, string? runID = null, string? firstExecutionRunID = null) { /*...*/ }
+
+    public WorkflowHandle<TResult> GetWorkflowHandle<TResult>(
+        string workflowID, string? runID = null, string? firstExecutionRunID = null) { /*...*/ }
+
+    // GetAsyncActivityHandle
+
+    public AsyncActivityHandle GetAsyncActivityHandle(string workflowID, string? runID, string activityID) { /*...*/ }
+
+    public AsyncActivityHandle GetAsyncActivityHandle(byte[] taskToken) { /*...*/ }
+
+#if NETCOREAPP3_0_OR_GREATER
+
+    // ListWorkflowsAsync
+
+    public WorkflowExecutionEnumerator ListWorkflows(
+        string query,
+        ListWorkflowsOptions? options = null,
+    ) { /*...*/ }
+
+#endif
+}
+```
+
+Notes:
+
+* There will be `ExecuteWorkflowAsync` helper methods for each `StartWorkflowAsync` overload there
+* It was decided forcing an options object is easier than lots of parameters for start workflow
+* Notice that there is a string-based workflow `StartWorkflowAsync` call that accepts a return type and one that doesn't
+  and returns a handle for `Nothing`. The user can specify a return type on `WorkflowHandle.GetResultAsync` if they
+  want instead of at start time (or just never altogether if the return doesn't matter to the user).
+* For string-based calls to `StartWorkflowAsync`, we are accepting all args at once via `object[]`. But since array
+  types are covariant in .NET, this means you could have a workflow that accepts `string[]` and forget that needs to
+  just be the first value of the object array.
+  * Cannot think of a good way to make this better. We could:
+    * Use varargs (i.e. `params`), but not having options as the last parameter is ugly DX
+    * Use some kind of forced wrapper, e.g. `IArgumentCollection`, but it's ugly too
+    * Accept args in options instead of as param, but that options set is reused for other calls
+    * Accept both single-arg and multi-arg w/ "unset" defaults, but that is confusing and most .NET devs use their
+      positional parameters unnamed
+    * Have separate overloads for no-arg, 1-arg, and multi-arg but that's not any clearer
+* For `GetWorkflowHandle`, we cannot accept a workflow reference because in .NET if you don't type all type arguments
+  (i.e. including that parameter type), you have to help the type inferencer out. So since they'd have to provide a type
+  argument anyways, we make it where they either can here or wait until `GetResultAsync` to do so.
```

Comment:
This is the saddest bit for me. Having to refer to the workflow _result_ type rather than the _workflow_ type feels like a nasty disconnect.

It seems like we could do what is done for signalling/querying and have it not be too ugly? Something like:

```csharp
var handle = await client.GetWorkflowHandle(IGreetingWorkflow.Ref.RunAsync, "my-workflow-id");
```

And now this can do the inference since it's getting the result type from the `RunAsync` function. Right?

The downside here is of course that saying `RunAsync` here feels a bit odd, but I know I'd prefer to do it this way instead of possibly getting the return type wrong.

Reply:
> It seems like we could do what is done for signalling/querying and have it not be too ugly? Something like:
> ```csharp
> var handle = await client.GetWorkflowHandle(IGreetingWorkflow.Ref.RunAsync, "my-workflow-id");
> ```

This is the conundrum I spoke about in Slack. I so wish we could do that. Say you have this signature for `GetWorkflowHandle` that looks like `StartWorkflowAsync` to support your snippet there:

```csharp
public WorkflowHandle<TResult> GetWorkflowHandle<T, TResult>(
    Func<T, Task<TResult>> workflow,
    string workflowID,
    string? runID = null,
    string? firstExecutionRunID = null
)
```

You get this error:

> error CS0411: The type arguments for method 'TemporalClient.GetWorkflowHandle<T, TResult>(Func<T, Task<TResult>>, string, string?, string?)' cannot be inferred from the usage. Try specifying the type arguments explicitly.

This is because, I assume due to type reification, .NET requires all type arguments to be inferencable. So your example changes to:

```csharp
var handle = await client.GetWorkflowHandle<GreetingInfo, string>(IGreetingWorkflow.Ref.RunAsync, "my-workflow-id");
```

And no, you can't specify just one type arg, you have to specify them all. And no, there is no `_` or `?` or `unknown` or similar placeholder for an that "arg" type param there. And no you can't give it some covariant like `object` (e.g. accept `Func<object, Task<TResult>>`) and expect it to work (you get errors like "Argument 1: cannot convert from 'method group' to 'Func<object, Task<string>>'" or at the least, force a type arg anyways).

This is just a known limitation in the .NET world. Their generic type inferencer just doesn't allow you to ignore types you don't provide. As one SO commenter states:

> As a rule, you're not going to have any luck with type inference when your method signature has fewer types of parameters than it has generic type parameters.

Meaning if you don't accept the types from the user in your signature, the inferencer won't be able to figure them out. I have spent a long time on this problem, enough to make me even consider switching to the proxy+lambda approach of Azure durable entities. But I figured:

* This only occurs on `GetWorkflowHandle` and, in the future, `GetExternalWorkflowHandle` on workflows
* People need a way to specify the return type anyways when using untyped handles
* I can provide an no-type-arg-required form for people that don't care about the return
* I can let people defer their type designation until they actually need the result (untyped when getting handle, specify result at result time)
* Other than the result type, there is no other benefit to accepting the workflow run method reference here. At least in Python I can force all signals/queries on the workflow handle to be of the same workflow class the run call is on, but not .NET.

After pouring over this forever, I believe what I have here is the best solution.

Reply:
OK, this is the part I was missing:
> Meaning if you don't accept the types from the user in your signature, the inferencer won't be able to figure them out. 

So signal works because one of the args is one of the generic types. Frustrating.

I think it'd still be nice if there was some static function, ex: `MyWorkflow.makeHandle(client)` which would do the write thing with types, but since I guess we don't control the whole interface/class, only certain functions, that doesn't quite work.

Reply:
> Frustrating

Exactly.

> that doesn't quite work.

Yeah, no matter how much I twist, I can't find a way given a workflow run method reference only, how to make a return-type-inferred handle from it if that run method has an argument.

Reply:
I don't really want to push in that direction, but of curiosity, would it works if workflows had to extend from a base interface, with the `RunAsync` method's return type being a generic parameter on the interface?

Reply:
Yes, if we force everyone to implement an interface that specifies the return type generically, then we can have functions that return a handle for that. So if you have:

```csharp
class MyWorkflow : IWorkflow<MyReturnType>
{
  public async Task<MyReturnType> RunAsync() { /*...*/ }
}
```

Then you can have:

```
public WorkflowHandle<TResult> GetWorkflowHandle<TResult>(IWorkflow<T> workflow...
```

But, you'd have to have a different interface for one that accepts an arg vs no arg (and to do it properly, a different one for no return too because there is no generic `void`/"unit" type in .NET)

Changing up the entire SDK to avoid some user specifying return type when they happen to manually get a handle is a bit much.